### PR TITLE
Modes

### PR DESCRIPTION
**Pull request by ebell495**
All five modes from SP 800-38A have been implemented. Also secure coding comments and practices have been implemented.

### CODE REVIEW

File: src/AESmodes.cpp [hunk 1]
```diff
@@ -5,268 +5,811 @@
...
+        const std::size_t plaintextLength = inputSize + padLength;
+
+        // Plaintext accommodates both the input and the necessary padding
+        std::vector<unsigned char> plaintext;
+        plaintext.reserve(plaintextLength);
+        plaintext = input;
+
+        // TODO: investigate whether PKCS#7 is the best choice for padding (padding oracle attack)
+        for (std::size_t i = 0; i < padLength; i++) {
+            // PKCS#7 padding (source: https://www.ibm.com/docs/en/zos/2.1.0?topic=rules-pkcs-padding-method)
```
Comment (rajrkane) [inline src/AESmodes.cpp:355]: Should ctr mode be using padding? I thought that's just for ecb cbc and cfb

Comment (ebell495) [reply src/AESmodes.cpp:355]: ECB, CBC and CFB require padding and the others don’t, but we can still use padding to figure out the size of the decrypted plaintext.

File: src/AESmodes.cpp [hunk 2]
```diff
@@ -5,268 +5,811 @@
...
+            // Secure coding: CTR50-CPP. Guarantee that container indices and iterators are within the valid range
+            output.push_back(outputBlock.at(j) ^ plaintext.at(j));
+        }
+
+        // Loop over the number of subsequent blocks
+        for (std::size_t i = 1; i < plaintextLength / NUM_BYTES; i++) {
+            // Loop over block size and fill each block
+            for (std::size_t j = 0; j < NUM_BYTES; j++) {
+                // Secure coding: CTR50-CPP. Guarantee that container indices and iterators are within the valid range
+                block.at(j) = output.at(j + ((i - 1) * NUM_BYTES));
```
Comment (rajrkane) [inline src/AESmodes.cpp:503]: This looks like 128-bit CFB mode. Could you confirm?

Comment (ebell495) [reply src/AESmodes.cpp:503]: Correct

**Review by rajrkane — APPROVED** [review]
Looks good to me!

File: src/AESmodes.cpp [hunk 3]
```diff
@@ -5,268 +5,811 @@
...
+            decrypt(block, outputPadded, key);
+
+            // Copy decrypted block to the output
+            for (std::size_t j = 0; j < NUM_BYTES; j++) {
+                // Secure coding: CTR50-CPP. Guarantee that container indices and iterators are within the valid range
+                output.push_back(outputPadded.at(j));
+            }
+        }
+
+        if (!remove_padding(output)) {
```
Comment (rajrkane) [inline src/AESmodes.cpp:133]: Wouldn't printing this allow a padding oracle attack? `remove_padding` returns `false` at the first incorrect byte, so the error message is also printed at the first incorrect byte.

Comment (rajrkane) [reply src/AESmodes.cpp:133]: Never mind, this is fine since the error message for padding and for a general decryption error is the same.

**Review by JacobR098 — APPROVED** [review]
Looks good!

**Pull request merged**